src/utils_request.js

```diff
@@ -0,0 +1,272 @@
+ *  similarly as `deflate` as `br` encodings.
+ * @param [options.json=false]
+ *  Sets body to JSON representation of value and adds `Content-type: application/json` header.
+ *  Additionally, parses the response body as JSON, i.e. the `body` property of the returned object
+ *  is the result of `JSON.parse()`. Throws an error if response cannot be parsed as JSON.
+ * @param [options.timeout]
+ *  Integer containing the number of milliseconds to wait for a server to send
+ *  response headers (and start the response body) before aborting the request.
+ *  Note that if the underlying TCP connection cannot be established, the OS-wide
+ *  TCP connection timeout will overrule the timeout option (the default in Linux can be anywhere from 20-120 seconds).
+ * @param [options.proxy]
+ *  An HTTP proxy to be used. Supports proxy authentication with Basic Auth.
+ * @param [options.strictSSL=true]
+ *  If `true`, requires SSL/TLS certificates to be valid.
+ * @param [options.abortFunction]
+ *  A function that determines whether the request should be aborted. It is called when the server
+ *  responds with the HTTP headers, but before the actual data is downloaded.
+ *  The function receives a single argument - an instance of Node's
+ *  [`http.IncomingMessage`](https://nodejs.org/api/http.html#http_class_http_incomingmessage)
+ *  class and it should return `true` if request should be aborted, or `false` otherwise.
+ * @param [options.throwOnHttpError=false]
+ *  If set to true function throws and error on 4XX and 5XX response codes.
+ * @return {{ response, body }}
+ *  Returns an object with two properties: `response` is the instance of
+ *  Node's [`http.IncomingMessage`](https://nodejs.org/api/http.html#http_class_http_incomingmessage) class,
+ *  `body` is a `String`, `Buffer` or `Object`, depending on the `encoding` and `json` options.
+ */
+export const requestBetter = async (options) => {
+    return new Promise((resolve, reject) => {
+        // Using the streaming API of Request to be able to
+        // handle the response based on headers receieved.
+        const defaultOptions = {
+            json: false,
+            gzip: false,
+            maxRedirects: 10,
+            followRedirect: true,
+            removeRefererHeader: false,
+            headers: {},
+            method: 'GET',
+            throwOnHttpError: false,
+            abortFunction: () => false,
+        };
+
+        const opts = Object.assign(defaultOptions, options);
+        const method = opts.method.toLowerCase();
+        const request = rqst[method](opts)
+            .on('error', err => reject(err))
+            .on('response', async (res) => {
+                const shouldAbort = opts.abortFunction(res);
+                if (shouldAbort) {
+                    request.abort();
+                }
+
+                // No need to catch invalid content header - it is already caught by request
+                const { type, encoding } = contentType.parse(res);
```

Comment:
What if the `request` and `contentType` validations are different? We still need to handle this because it's an error in a callback and it could kill the process.

Reply:
You are right. I thought that the request package uses the content-type package, but I was wrong. The error that put me on the wrong path was coming from express.